components/PlaybackTask.brs

```diff
@@ -0,0 +1,253 @@
+    m.skipAds = true
+  else if arrayUtils_includes(startPlaybackEvents, eventType)
+    startPlayback()
+  else if eventType = types.USERCANCELSTREAM
+    exitPlayback()
+  end if
+end sub
+
+function getPreroll() as Object
+  ads = m.raf.getAds()
+  if ads = invalid then return false
+  
+  result = invalid
+  for each adPod in ads
+    if adPod.rendersequence = "preroll"
+      result = adPod
+      exit for
+    end if
+  end for
+
+  return result
+end function
+
+function handleAds(ads) as Boolean
+  resumePlayback = true
+
+  if ads <> invalid AND ads.ads.count() > 0
+    m.currentAdPod = ads
+    firstAd = ads.ads[0] 'Assume truex can only be first ad
+
+    if firstAd.adParameters <> invalid
+      m.truexAd = firstAd
+      m.truexAd.adParameters = parseJSON(m.truexAd.adParameters)
+      m.truexAd.renderSequence = ads.renderSequence
+      ads.ads.delete(0) ' Removes it from certain use cases due to raf
+
+      playTrueXAd()
+      resumePlayback = false
+    else ' Non-TrueX ads
+      hidePlayback()
+      watchedAd = m.raf.showAds(ads, invalid, m.adFacade) 'Takes thread ownership until complete or exit
+
+      if watchedAd
+        resumePlayback = true 
+      else
+        resumePlayback = false
+        exitPlayback() 
+      end if
+    end if
+  end if
+
+  return resumePlayback
+end function
+
+sub playTrueXAd()
+    ? "TRUE[X] >>> PlaybackTask::playTrueXAd() - instantiating TruexAdRenderer ComponentLibrary..."
+
+    ' instantiate TruexAdRenderer and register for event updates
+    m.adRenderer = m.top.adFacade.createChild("TruexLibrary:TruexAdRenderer")
+    m.adRenderer.observeFieldScoped("event", m.port)
+
+    ' use the companion ad data to initialize the true[X] renderer
+    tarInitAction = {
+      type: "init",
+      adParameters: m.truexAd.adParameters,
+      supportsUserCancelStream: true, ' enables cancelStream event types, disable if Channel does not support
+      slotType: ucase(m.truexAd.rendersequence),
+      logLevel: 1, ' Optional parameter, set the verbosity of true[X] logging, from 0 (mute) to 5 (verbose), defaults to 5
+      channelWidth: 1920, ' Optional parameter, set the width in pixels of the channel's interface, defaults to 1920
+      channelHeight: 1080 ' Optional parameter, set the height in pixels of the channel's interface, defaults to 1080
+    }
+    ? "TRUE[X] >>> PlaybackTask::playTrueXAd() - initializing TruexAdRenderer with action=";tarInitAction
+    m.adRenderer.action = tarInitAction
+
+    hidePlayback()
+
+    ? "TRUE[X] >>> PlaybackTask::playTrueXAd() - starting TruexAdRenderer..."
+    m.adRenderer.action = { type: "start" }
+    m.adRenderer.focusable = true
+    m.adRenderer.SetFocus(true)
+end sub
+
+sub startPlayback()
+    cleanUpAdRenderer()
+
+    if m.skipAds AND m.currentAdPod <> invalid then
+      m.currentAdPod.viewed = true
+      m.currentAdPod = invalid
+    end if
+
+    ' Check if we need to play other (non-truex) ads
+    play = handleAds(m.currentAdPod)
```

Comment:
I was a bit conflicted with the name because this is also responsible for starting playback the first time.  Resume makes less sense for that initial case.  It didn't seem necessary to create a separate start playback flow when it is handled by one general flow.